File: tap/di/cmap/cmap.py
from collections import namedtuple
from math import inf, sqrt

Point = namedtuple('Point', ('x', 'y'))


def read_point(f):
    x, y = map(int, next(f).split())
    return Point(x, y)


def distance_square(a, b):
    return (a.x - b.x) ** 2 + (a.y - b.y) ** 2


def min_distance(xs, ys):
    if len(xs) <= 1:
        return inf
    if len(xs) == 2:
        return distance_square(xs[0], xs[1])
    if len(xs) == 3:
        return min(distance_square(xs[0], xs[1]),
                   distance_square(xs[0], xs[2]),
                   distance_square(xs[1], xs[2]))

    # print('Sortate după X:', *xs)
    # print('Sortate după Y:', *ys)

    mid = len(xs) // 2
    mid_x = xs[mid].x
    # print(f'Despart punctele prin dreapta x={mid_x}')

    xs_left, xs_right = xs[:mid], xs[mid:]
    ys_left, ys_right = [], []
    for pt in ys:
        if pt.x < mid_x:
            ys_left.append(pt)
        else:
            ys_right.append(pt)

    # print(*xs_left, '<', *xs_right)
    # print(*ys_left, '<', *ys_right)

    min_left = min_distance(xs_left, ys_left)
    min_right = min_distance(xs_right, ys_right)

    d = min(min_left, min_right)

    closer = [pt for pt in ys if (pt.x - mid_x) ** 2 < d]
    # print(*closer)
    i = 0
    while i < len(closer) - 1:
        j = i + 1
        while j < len(closer) and j - i <= 8:
            d = min(d, distance_square(closer[i], closer[j]))
            j += 1
        i += 1

    return min(min_left, min_right)


with open('cmap.in', 'r') as fin:
    num_points = int(next(fin))
    points = [read_point(fin) for _ in range(num_points)]

sorted_by_x = sorted(points, key=lambda pt: pt.x)
sorted_by_y = sorted(points, key=lambda pt: pt.y)

min_dist = min_distance(sorted_by_x, sorted_by_y)
min_dist = sqrt(min_dist)

with open('cmap.out', 'w') as fout:
    print(min_dist, file=fout)

File: tap/di/cmap/cmap.cpp
#include <bits/stdc++.h>
using namespace std;

struct Point {
    int x, y;
};

istream& operator>>(istream& is, Point& pt) {
    return is >> pt.x >> pt.y;
}

ostream& operator<<(ostream& os, const Point& pt) {
    return os << "(" << pt.x << ", " << pt.y << ")";
}

double distance_square(Point a, Point b) {
    double dx = double(a.x) - double(b.x),
        dy = double(a.y) - double(b.y);

    return dx * dx + dy * dy;
}

double min_distance(size_t num_pts, const Point* xs, const Point* ys) {
    if (num_pts <= 1) {
        return INFINITY;
    }
    if (num_pts == 2) {
        return distance_square(xs[0], xs[1]);
    }
    if (num_pts == 3) {
        double a = distance_square(xs[0], xs[1]),
            b = distance_square(xs[0], xs[2]),
            c = distance_square(xs[1], xs[2]);
        return min(a, min(b, c));
    }

    int mid = num_pts / 2;
    int mid_x = xs[mid].x;

    const Point* left_xs = xs;
    const Point* right_xs = xs + mid;
    vector<Point> left_ys, right_ys;
    left_ys.reserve(mid);
    right_ys.reserve(mid);

    for (size_t i = 0; i < num_pts; ++i) {
        if (ys[i].x < mid_x) {
            left_ys.push_back(ys[i]);
        } else {
            right_ys.push_back(ys[i]);
        }
    }

    double min_left = min_distance(mid, left_xs, left_ys.data()),
        min_right = min_distance(mid, right_xs, right_ys.data());

    double d = min(min_left, min_right);

    vector<Point> closer;
    closer.reserve(num_pts);
    for (size_t i = 0; i < num_pts; ++i) {
        double dx = abs(ys[i].x - mid_x);
        if (dx <= d) {
            closer.push_back(ys[i]);
        }
    }

    if (closer.size() > 0) {
        size_t i = 0;
        while (i < closer.size() - 1) {
            size_t j = i + 1;
            while (j < closer.size() && (j - i) <= 8) {
                d = min(d, distance_square(closer[i], closer[j]));
                ++j;
            }
            ++i;
        }
    }

    return d;
}

int main() {
    ifstream in("cmap.in");

    size_t num_pts;
    in >> num_pts;

    vector<Point> pts(num_pts);
    for (size_t i = 0; i < num_pts; ++i) {
        in >> pts[i];
    }

    vector<Point> xs = pts, ys = pts;

    sort(xs.begin(), xs.end(), [] (const Point& a, const Point& b) { return a.x < b.x; });
    sort(ys.begin(), ys.end(), [] (const Point& a, const Point& b) { return a.y < b.y; });

    ofstream out("cmap.out");
    double d_squared = min_distance(num_pts, xs.data(), ys.data());
    out << fixed << setprecision(6) << sqrt(d_squared) << '\n';
}
